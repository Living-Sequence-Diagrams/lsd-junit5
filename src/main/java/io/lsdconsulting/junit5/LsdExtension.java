package io.lsdconsulting.junit5;

import com.lsd.core.IdGenerator;
import com.lsd.core.LsdContext;
import com.lsd.core.properties.LsdProperties;
import j2html.tags.UnescapedText;
import org.junit.jupiter.api.extension.AfterAllCallback;
import org.junit.jupiter.api.extension.AfterTestExecutionCallback;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.jupiter.api.extension.TestWatcher;
import org.junit.platform.commons.util.ExceptionUtils;

import java.lang.annotation.Annotation;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Optional;
import java.util.function.Consumer;

import static com.lsd.core.domain.Status.*;
import static com.lsd.core.report.PopupContent.popupDiv;
import static j2html.TagCreator.*;
import static java.util.Arrays.stream;
import static java.util.function.Predicate.not;
import static org.junit.platform.commons.util.StringUtils.isBlank;

/**
 * Junit 5 extension to create LSD reports based on test results.
 * (Not that this extension does not capture any events for the LSD diagrams)
 */
public class LsdExtension implements TestWatcher, AfterTestExecutionCallback, AfterAllCallback {

    private final LsdContext lsdContext = LsdContext.getInstance();
    private final IdGenerator idGenerator = lsdContext.getIdGenerator();
    private final boolean hideStacktrace = LsdProperties.getBoolean("lsd.junit5.hideStacktrace");

    @Override
    public void testSuccessful(ExtensionContext context) {
        lsdContext.completeScenario(prefixParentDisplayName(context), p(
                h4().with(new UnescapedText("&#127808; Test Passed")).withClass("success")
        ).render(), SUCCESS);
    }

    @Override
    public void testDisabled(ExtensionContext context, Optional<String> reason) {
        lsdContext.completeScenario(prefixParentDisplayName(context), p(
                h4("Test Disabled").withClass("warn")
        ).render(), FAILURE);
    }

    @Override
    public void testAborted(ExtensionContext context, Throwable cause) {
        var description = createErrorDescription(cause, "Test Aborted!");
        lsdContext.completeScenario(prefixParentDisplayName(context), description, FAILURE);
    }

    @Override
    public void testFailed(ExtensionContext context, Throwable cause) {
        var description = createErrorDescription(cause, "&#10060; Failed!");
        lsdContext.completeScenario(prefixParentDisplayName(context), description, ERROR);
    }

    @Override
    public void afterTestExecution(ExtensionContext context) {
        Object testInstance = context.getRequiredTestInstance();
        additionalProcessing(testInstance, LsdPostTestProcessing.class);
    }

    @Override
    public void afterAll(ExtensionContext context) {
        if (isNested(context)) {
            return;
        }
        lsdContext.completeReport(context.getDisplayName());
        lsdContext.createIndex();
    }

    private boolean isNested(ExtensionContext context) {
        return context.getParent()
                .map(ExtensionContext::getParent)
                .map(Optional::isPresent)
                .orElse(false);
    }

    private String prefixParentDisplayName(ExtensionContext context) {
        var parent = context.getParent();
        if (parent.isPresent()) {
            var parentDisplayName = prefixParentDisplayName(parent.get());
            var separator = isBlank(parentDisplayName) ? "" : ": ";
            return parentDisplayName + separator + context.getDisplayName();
        }
        return "";
    }

    private String createErrorDescription(Throwable cause, String header) {
        var contentId = idGenerator.next();
        String exceptionMessage = extractMessage(cause);
        return p(
                h4().with(new UnescapedText(header)).withClass("error"),
                a().withHref("#" + contentId).withText(exceptionMessage),
                popupDiv(contentId, "Stacktrace", readStackTrace(cause))
        ).render();
    }

    private void additionalProcessing(final Object instance, Class<? extends Annotation> annotation) {
        Class<?> klass = instance.getClass();
        while (klass != Object.class) {
            stream(klass.getDeclaredMethods())
                    .filter(method -> method.isAnnotationPresent(annotation))
                    .forEach(invokeMethodOn(instance));
            klass = klass.getSuperclass();
        }
    }

    private String extractMessage(Throwable cause) {
        return Optional.ofNullable(cause)
                .map(Throwable::getMessage)
                .orElse("");
    }

    private String readStackTrace(Throwable cause) {
        return Optional.ofNullable(cause)
                .filter(not(throwable -> hideStacktrace))
                .map(ExceptionUtils::readStackTrace)
                .orElse("[Displaying the stacktrace was disabled or no cause was provided]");
    }

    private Consumer<Method> invokeMethodOn(Object instance) {
        return method -> {
            try {
                method.setAccessible(true);
                method.invoke(instance);
            } catch (IllegalAccessException | InvocationTargetException e) {
                throw new RuntimeException(e);
            }
        };
    }
}