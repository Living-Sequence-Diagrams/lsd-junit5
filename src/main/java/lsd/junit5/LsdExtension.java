package lsd.junit5;

import com.lsd.IdGenerator;
import com.lsd.LsdContext;
import com.lsd.properties.LsdProperties;
import j2html.tags.UnescapedText;
import org.junit.jupiter.api.extension.AfterAllCallback;
import org.junit.jupiter.api.extension.AfterTestExecutionCallback;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.jupiter.api.extension.TestWatcher;
import org.junit.platform.commons.util.ExceptionUtils;

import java.lang.annotation.Annotation;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Optional;
import java.util.function.Consumer;

import static com.lsd.OutcomeStatus.*;
import static com.lsd.report.model.PopupContent.popupDiv;
import static j2html.TagCreator.*;
import static java.util.Arrays.stream;
import static java.util.function.Predicate.not;
import static org.junit.platform.commons.util.StringUtils.isBlank;

public class LsdExtension implements TestWatcher, AfterTestExecutionCallback, AfterAllCallback {

    private final LsdContext lsdContext = LsdContext.getInstance();
    private final IdGenerator idGenerator = lsdContext.getIdGenerator();
    private final boolean hideStacktrace = LsdProperties.getBoolean("lsd.junit5.hideStacktrace");

    @Override
    public void testSuccessful(ExtensionContext context) {
        lsdContext.completeScenario(prefixParentDisplayName(context), p(
                h4().with(new UnescapedText("&#127808; Test Passed")).withClass(SUCCESS.getCssClass())
        ).render(), SUCCESS);
    }

    @Override
    public void testDisabled(ExtensionContext context, Optional<String> reason) {
        lsdContext.completeScenario(prefixParentDisplayName(context), p(
                h4("Test Disabled").withClass(WARN.getCssClass())
        ).render(), WARN);
    }

    @Override
    public void testAborted(ExtensionContext context, Throwable cause) {
        var description = createErrorDescription(cause, "Test Aborted!");
        lsdContext.completeScenario(prefixParentDisplayName(context), description, WARN);
    }

    @Override
    public void testFailed(ExtensionContext context, Throwable cause) {
        var description = createErrorDescription(cause, "&#10060; Failed!");
        lsdContext.completeScenario(prefixParentDisplayName(context), description, ERROR);
    }

    @Override
    public void afterTestExecution(ExtensionContext context) {
        Object testInstance = context.getRequiredTestInstance();
        additionalProcessing(testInstance, LsdPostTestProcessing.class);
    }

    @Override
    public void afterAll(ExtensionContext context) {
        if (notNestedTest(context)) {
            lsdContext.completeReport(context.getDisplayName());
            lsdContext.createIndex();
        }
    }

    private boolean notNestedTest(ExtensionContext context) {
        return !context.getParent()
                .map(ExtensionContext::getParent)
                .map(Optional::isPresent)
                .orElse(true);
    }

    private String prefixParentDisplayName(ExtensionContext context) {
        var parent = context.getParent();
        if (parent.isPresent()) {
            var parentDisplayName = prefixParentDisplayName(parent.get());
            var separator = isBlank(parentDisplayName) ? "" : ": ";
            return parentDisplayName + separator + context.getDisplayName();
        }
        return "";
    }

    private String createErrorDescription(Throwable cause, String header) {
        var contentId = idGenerator.next();
        String exceptionMessage = extractMessage(cause);
        return p(
                h4().with(new UnescapedText(header)).withClass(ERROR.getCssClass()),
                a().withHref("#" + contentId).withText(exceptionMessage),
                popupDiv(contentId, "Stacktrace", readStackTrace(cause))
        ).render();
    }

    private void additionalProcessing(final Object instance, Class<? extends Annotation> annotation) {
        Class<?> klass = instance.getClass();
        while (klass != Object.class) {
            stream(klass.getDeclaredMethods())
                    .filter(method -> method.isAnnotationPresent(annotation))
                    .forEach(invokeMethodOn(instance));
            klass = klass.getSuperclass();
        }
    }

    private String extractMessage(Throwable cause) {
        return Optional.ofNullable(cause)
                .map(Throwable::getMessage)
                .orElse("");
    }

    private String readStackTrace(Throwable cause) {
        return Optional.ofNullable(cause)
                .filter(not(throwable -> hideStacktrace))
                .map(ExceptionUtils::readStackTrace)
                .orElse("[Displaying the stacktrace was disabled or no cause was provided]");
    }

    private Consumer<Method> invokeMethodOn(Object instance) {
        return method -> {
            try {
                method.setAccessible(true);
                method.invoke(instance);
            } catch (IllegalAccessException | InvocationTargetException e) {
                throw new RuntimeException(e);
            }
        };
    }
}