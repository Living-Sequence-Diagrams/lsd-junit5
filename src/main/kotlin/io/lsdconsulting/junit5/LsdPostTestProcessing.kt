package io.lsdconsulting.junit5

/**
 * Test that use the `LsdExtension` to generate LSD reports can use this annotation to flag a
 * method (can be private but must take no arguments) that should be invoked after each Test has run but before the
 * sequence diagram is generated.
 *
 *
 * This may be useful if it is necessary to add additional interactions or interesting givens to the TestState instance
 * before the sequence diagrams are generated.
 */
@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
annotation class LsdPostTestProcessing  