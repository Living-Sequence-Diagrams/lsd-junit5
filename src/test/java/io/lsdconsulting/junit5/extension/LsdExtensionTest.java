package io.lsdconsulting.junit5.extension;

import com.lsd.core.LsdContext;
import io.lsdconsulting.junit5.LsdExtension;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtensionContext;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.Optional;

import static com.lsd.core.properties.LsdProperties.OUTPUT_DIR;
import static com.lsd.core.properties.LsdProperties.get;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class LsdExtensionTest {
    private final ExtensionContext mockTestContext = mock(ExtensionContext.class);
    private final LsdContext lsdContext = LsdContext.getInstance();
    private final LsdExtension lsdExtension = new LsdExtension();
    private final File indexFile = new File(get(OUTPUT_DIR), "lsd-index.html");

    @BeforeEach
    public void setup() {
        lsdContext.clear();
    }

    @Test
    void approvalTestForTestFailedCallback() {
        configureMockContextToReturnDisplayNames();

        lsdExtension.testFailed(mockTestContext, new AssertionError("Expected <true> but was <false>"));

        Approvals.verify(lsdContext.completeReport("Failure Test Report").toFile());
    }

    @Test
    void approvalTestForTestSuccessfulCallback() throws IOException {
        configureMockContextToReturnDisplayNames();

        lsdExtension.testSuccessful(mockTestContext);

        Approvals.verify(copyOfFile(lsdContext.completeReport("Successful Test Report")));
    }

    @Test
    void approvalTestForTestAbortedCallback() throws IOException {
        configureMockContextToReturnDisplayNames();

        lsdExtension.testAborted(mockTestContext, new RuntimeException("Aborted for testing reasons"));

        Approvals.verify(copyOfFile(lsdContext.completeReport("Aborted Test Report")));
    }

    @Test
    void approvalTestForTestDisabledCallback() throws IOException {
        configureMockContextToReturnDisplayNames();

        lsdExtension.testDisabled(mockTestContext, Optional.of("Disabled reason"));

        Approvals.verify(copyOfFile(lsdContext.completeReport("Disabled Test Report")));
    }

    @Test
    void generateIndexAfterAllTestsInTestClass() {
        indexFile.delete();
        when(mockTestContext.getParent()).thenReturn(Optional.empty());
        when(mockTestContext.getDisplayName()).thenReturn("a display name");

        lsdExtension.afterAll(mockTestContext);

        assertThat(indexFile).exists();
    }

    @Test
    void doesNotGenerateIndexWhenTestIsNested() {
        indexFile.delete();
        when(mockTestContext.getParent()).thenReturn(Optional.of(mockTestContext));
        when(mockTestContext.getDisplayName()).thenReturn("a nested display name");

        lsdExtension.afterAll(mockTestContext);

        assertThat(indexFile).doesNotExist();
    }

    private void configureMockContextToReturnDisplayNames() {
        when(mockTestContext.getRequiredTestInstance()).thenReturn(this);
        when(mockTestContext.getDisplayName()).thenReturn("The test name");
        when(mockTestContext.getParent())
                .thenReturn(Optional.of(mockTestContext))
                .thenReturn(Optional.empty());
    }

    private File copyOfFile(Path original) throws IOException {
        Path copy = Path.of(original.getParent().toString(), "copy_" + original.getFileName());
        Files.copy(original, copy, StandardCopyOption.REPLACE_EXISTING);
        return copy.toFile();
    }
}
