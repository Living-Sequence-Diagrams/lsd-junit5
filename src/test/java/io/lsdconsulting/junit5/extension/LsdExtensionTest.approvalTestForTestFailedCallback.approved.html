<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns="http://www.w3.org/1999/html" xml:lang="en" lang="en">
<head>
    <title>Failure Test Report</title>
    <link rel="stylesheet" href="style.css">
</head>
<body>
<main role="main">
    <span onclick="location.href='lsd-index.html';" title="back to index page" class="back_to_index"></span>
    <h1 class="logo">LSD</h1>
    <h1>Failure Test Report</h1>
    
    
    

    <article id="2">
        <span onclick="location.href='#top';" title="top of page" class="back_to_top"></span>
        <h2 class="error">The test name</h2>
        <section class="description error">
            <h3 class="error">Description</h3>
            <p><h4 class="error">&#10060; Failed!</h4><a href="#1">Expected &lt;true&gt; but was &lt;false&gt;</a><div id="1" class="overlay" onclick="location.href=&#x27;#!&#x27;;"><div class="popup" onclick="event.stopPropagation();"><h2>Stacktrace</h2><a class="close" href="#!">&times;</a><div class="content">[Displaying the stacktrace was disabled or no cause was provided]</div></div></div></p>
        </section>
        
        
    </article>
    
</main>
</body>
<footer>
    <script type = "text/javascript" src = "https://d3js.org/d3.v4.min.js"></script>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/highlightjs/cdn-release@11.0.0/build/styles/default.min.css">
    <script src="custom.js"></script>
    <script language = "javascript">
        addSliderForSvgZoom();
        highlightLifelinesWhenClicked();
    </script>
    <script src="https://cdn.jsdelivr.net/gh/highlightjs/cdn-release@11.0.0/build/highlight.min.js"></script>
    <script>hljs.highlightAll();</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js"></script>
    <script>
        highlightKeywords();
        
            
        
        unMarkSvg();
    </script>
</footer>
</html>
